### PR TITLE
Feature/hip 30 update the api server to not be main thread blocking

### PR DESCRIPTION
**Pull request by mateimol**
Update the files such that the server is not main thread blocking.

### CODE REVIEW

File: hippo/server.py [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+    is to be able to launch the uvicorn server in a separate thread. The purpose is to not block the main thread.
+
+    The class is based on the comments found at this address:
+    https://github.com/encode/uvicorn/issues/742#issuecomment-674411676
+
+    The class inherits from the uvicorn.Server class.
+    """
+
+    def __init__(self, app: FastAPI, host: str = "localhost", port: int = 8000, log_level: str = "info") -> None:
+        """[summary]
```
Comment (yribeiro) [inline hippo/server.py]: Missing docstring.

File: hippo/launcher.py [hunk 2]
```diff
@@ -84,10 +89,27 @@ def app(self) -> FastAPI:
...
+        self.__server.run_thread()
+
+    def stop(self) -> None:
+        """
+        Method to stop the apps uvicorn server.
         """
-        uvicorn.run(app=self.__app, host=self.__host, port=self.__port)
+        self.__server.stop_thread()
+
+    @contextlib.contextmanager
```
Comment (yribeiro) [inline hippo/launcher.py:105]: Typically I would expect this to be run with enter and exit methods.

File: hippo/server.py [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+            host (str, optional): IP address host to launch service on.. Defaults to 'localhost'.
+            port (int, optional): Port number to host service on. Defaults to 8000.
+            log_level (str, optional):The level of logging we want the server to have. Defaults to 'info'.
+
+        Raises:
+            RuntimeError: [description]
+        """
+        assert isinstance(host, str), f"Expected type string for arg: host, got type: {type(host)}"
+        assert isinstance(port, int), f"Expected type int for arg: port, got type: {type(port)}"
+        # assert isinstance(app, FastAPI), f"Expected type FastAPI for arg: app, got type: {type(app)}"
```
Comment (yribeiro) [inline hippo/server.py]: Lingering commented line. Please remove.

File: hippo/server.py [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+
+        super().__init__(
+            self.__config
+        )  # not entirely sure what this does, but I assume it has to do with the inheritance of uvicorn.Server
+
+        # define the thread as a class property/parameter so that the "run_in_thread" and "stop_thread" can be run on the same thread.
+        self.__thread = threading.Thread(target=self.run, daemon=True)
+
+    # not sure what this function does either but it was in the tutorial that I saw
+    def install_signal_handlers(self) -> None:
```
Comment (yribeiro) [inline hippo/server.py]: This is for handling exit conditions from the terminal. Can be removed I think? Unless the base class needs it?

File: hippo/server.py [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+        pass
+
+    def run_thread(self) -> None:
```
Comment (yribeiro) [inline hippo/server.py]: Just `run()` should suffice?

File: hippo/server.py [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+        """Method for starting the thread containing the uvicorn server"""
+        self.__thread.start()
+
+    def stop_thread(self) -> None:
```
Comment (yribeiro) [inline hippo/server.py]: Just `stop()` should suffice?

File: pyproject.toml [hunk 7]
```diff
@@ -16,6 +16,7 @@ torch = "^1.10.0"
 Flask = "^2.0.2"
 fastapi = "^0.71.0"
 uvicorn = "^0.17.0"
+threaded = "^4.1.0"
```
Comment (yribeiro) [inline pyproject.toml:19]: This is threaded?

File: tests/test_server.py [hunk 8]
```diff
@@ -0,0 +1,44 @@
...
+            _ = Server(app=mockFastApi, host=test_host, port=test_port)
+
+    @patch("hippo.server.threading.Thread")
+    @patch("hippo.server.uvicorn.Config")
+    @patch("hippo.server.uvicorn.Server.run")
+    def test_server_initialises_config_and_thread_as_expected(
+        self, mockUvicornServer: Mock, mockUvicorn: Mock, mockThread: Mock
+    ) -> None:
+
+        Server(app="hippo.main:DEVAPP", host="localhost", port=12345, log_level="info")
```
Comment (yribeiro) [inline tests/test_server.py]: You traditionally have tests in sections like this: 

```python
# setup

# execute

# assert
```

or sometimes like this

```python
# setup

# execute and assert
````

**Review by yribeiro — CHANGES_REQUESTED** [review]

File: hippo/launcher.py [hunk 9]
```diff
@@ -33,6 +36,8 @@ def __init__(self, host: str, port: int) -> None:
         self.__host = host
         self.__port = port
         self.__app = self._build_fast_api_app()
+        # An intance of the Server class, this is an abstraction of the uvicorn server to allow the starting and stopping of the server in a separate thread.
```
Comment (yribeiro) [inline hippo/launcher.py]: This is strictly not an abstraction, it is a wrapper class.

File: hippo/server.py [hunk 10]
```diff
@@ -47,13 +47,17 @@ def run_thread(self) -> None:
         """Method for starting the thread containing the uvicorn server"""
         self.__thread.start()
 
-    def stop_thread(self) -> None:
-        """Method for stopping the thread containing the uvicorn server"""
+    def stop_thread(self, timeout: int = 5) -> None:
```
Comment (yribeiro) [inline hippo/server.py]: Just `run` and `stop` should suffice?

**Review by yribeiro — CHANGES_REQUESTED** [review]

File: hippo/server.py (hunk 10 above)
Comment (mateimol) [reply hippo/server.py]: done

File: hippo/server.py (hunk 1 above)
Comment (mateimol) [reply hippo/server.py]: I added Initialize

File: hippo/server.py (hunk 4 above)
Comment (mateimol) [reply hippo/server.py]: I removed it and all still works.

File: hippo/server.py (hunk 5 above)
Comment (mateimol) [reply hippo/server.py]: done

File: hippo/server.py (hunk 6 above)
Comment (mateimol) [reply hippo/server.py]: done

File: pyproject.toml (hunk 7 above)
Comment (mateimol) [reply pyproject.toml:19]: Yes, this is the threading library used for preventing the launcher from blocking the main thread.

File: tests/test_server.py (hunk 8 above)
Comment (mateimol) [reply tests/test_server.py]: done.

File: hippo/launcher.py (hunk 9 above)
Comment (mateimol) [reply hippo/launcher.py]: Modified the comment

**Review by yribeiro — APPROVED** [review]

**Pull request merged**